### ISSUE TITLE
How to continue when a grub installation fails? And why is it failing? :cry: 

### ISSUE THREAD

**Issue by doronbehar** [numtide/nixos-anywhere#120]
I've been trying our different configurations for about half a day and none of them work. It's either that I get errors such as:

```
updating GRUB 2 menu...
installing the GRUB 2 boot loader on /dev/sda...
Installing for i386-pc platform.
/nix/store/kps2xxdg43x7wkjxh97wp6zpqx2gzdqy-grub-2.06/sbin/grub-install: warning: this GPT partition label contains no BIOS Boot Partition; embedding won't be possible.
/nix/store/kps2xxdg43x7wkjxh97wp6zpqx2gzdqy-grub-2.06/sbin/grub-install: warning: Embedding is not possible.  GRUB can only be installed in this setup by using blocklists.  However, blocklists are UNRELIABLE and their use is discouraged..
/nix/store/kps2xxdg43x7wkjxh97wp6zpqx2gzdqy-grub-2.06/sbin/grub-install: error: will not proceed with blocklists.
/nix/store/mpli8kczf4x9j9icfr5gjgl0a9y7rybx-install-grub.pl: installation of GRUB on /dev/sda failed: No such file or directory
```

Or that I got a machine that is incapable of boot.

Each time this fails I ask my VPS provider to reinstall their default OS distro, and then I `sudo passwd` and give root ssh access for `nixos-anywhere`.

Here's the configuration that produced the errors above:

```nix
let
    device = "/dev/sda";
in {
  disko.devices = {
    disk = {
      "${builtins.baseNameOf device}" = {
        inherit device;
        type = "disk";
        content = {
          type = "table";
          format = "gpt";
          partitions = [
            {
              name = "ESP";
              start = "1MiB";
              end = "100MiB";
              bootable = true;
              content = {
                type = "filesystem";
                format = "vfat";
                mountpoint = "/boot";
              };
            }
            {
              name = "root";
              start = "100MiB";
              end = "-1G";
              part-type = "primary";
              content = {
                type = "filesystem";
                format = "ext4";
                mountpoint = "/";
              };
            }
            {
              name = "swap";
              start = "-1G";
              end = "100%";
              part-type = "primary";
              content = {
                type = "swap";
                randomEncryption = true;
              };
            }
          ];
        };
      };
    };
  };

  boot = {
    kernelPackages = pkgs.linuxPackages_latest;
    loader.grub = {
      enable = true;
      inherit device;
    };
  };
  # Other settings ...
}
```

Help will be appreciated!

**Comment by phaer** [comment]
It's trying to install MBR (non-EFI) boot, but you lack a partition fit for "embedding" grub.

You can either set `boot.loader.grub.devices = [ "nodev" ];` to disable the MBR thing and run GRUB with UEFI only, or create a small bios_grub partition as i.e. in https://github.com/nix-community/disko/blob/master/example/gpt-bios-compat.nix#L17

**Comment by doronbehar** [comment]
@phaer I've read all the simple examples at https://github.com/nix-community/disko/tree/master/example/ and they don't include grub configurations, so I wasn't sure how to take them and what grub configuration will fit them. Would it be possible please to get a full configuration that's working? I've tried at least 8-9 options and I'm getting frustrated...

**Comment by phaer** [comment]
@doronbehar Yes, disko is a partitioning tool and has little to do with bootloaders per se, but the standalone example does even include a grub example config. It's just a legacy/BIOS boot example, not a UEFI one.

Did you try my suggestion with your config above?

**Comment by doronbehar** [comment]
It'd have been ideal if you have provided a flake that would allow to run `nixos-anywhere` like this:

```
nix run github:numtide/nixos-anywhere -- --flake github:numtide/nixos-anywhere-examples\#example/simple-with-swap root@vps
```

Or:

```
nix run github:numtide/nixos-anywhere -- --flake github:numtide/nixos-anywhere\#example/simple-without-swap root@vps
```

This will help frustrated users like me to get something working when nothing works for them...

**Comment by phaer** [comment]
> I tried using grub.devices = [ "nodev" ]; and I removed grub.device and at least installation didn't fail. However it won't boot:

You are trying to boot an UEFI-only configuration with a normal BIOS here, that won't work. You either need a config like in the standalone example or get your host to boot in UEFI mode.

> If you are talking about the one from the README

I am talking about the one in the "stand-alone" directory inside the examples directory you've linked https://github.com/nix-community/disko/tree/master/example/stand-alone

> I realized that I need to replace type = "partition" with part-type = "partition" - outdated README, and then I got:

`part-type = "primary"` should be correct, but also the default. You should be able to just remove `type` from the partitions (but leave it for the table and the filesystems)

Regarding your last error, not sure whats wrong here at a glance, device numbering seems to be off. An fdisk -l or similar of that disk would be interesting. 

> It'd have been ideal if you have provided a flake that would allow to run nixos-anywhere like this:

I understand that this can be a complex and taunting topic, you probably wouldn't believe how many tries I had to give it while working just on a few patches to disko. But please do remember that this is all free software, written (mostly) by volunteers in their free time - you a welcome to contribute :) 
The problem with your example is that nixos-anywhere and disko are quite flexible with what the allow and that therefore a shared definition of "simple" is hard to achieve. Should "simple" support UEFI only? Or hybrid? Should it be ext4fs or btrfs? Grub or systemd-boot?, etc.

**Comment by doronbehar** [comment]
> I understand that this can be a complex and taunting topic, you probably wouldn't believe how many tries I had to give it while working just on a few patches to disko.

That's compelling :).

> But please do remember that this is all free software, written (mostly) by volunteers in their free time - you a welcome to contribute :)
The problem with your example is that nixos-anywhere and disko are quite flexible with what the allow and that therefore a shared definition of "simple" is hard to achieve. Should "simple" support UEFI only? Or hybrid? Should it be ext4fs or btrfs? Grub or systemd-boot?

Perhaps the naming should be different, but in general it'd be nice to have more examples that one can try each via `nix run`. I would contribute my configuration happily once I get it working!

> > If you are talking about the one from the README
> 
> I am talking about the one in the "stand-alone" directory inside the examples directory you've linked [nix-community/disko@`master`/example/stand-alone](https://github.com/nix-community/disko/tree/master/example/stand-alone?rgh-link-date=2023-05-31T09%3A19%3A28Z)
> 
> > I realized that I need to replace type = "partition" with part-type = "partition" - outdated README, and then I got:
> 
> `part-type = "primary"` should be correct, but also the default. You should be able to just remove `type` from the partitions (but leave it for the table and the filesystems)

I tried:

```nix
diskoModule:
{ config, pkgs, options, lib, ... }:

let
    device = "/dev/sda";
in {
  disko.devices = {
    disk = {
      "${builtins.baseNameOf device}" = {
        inherit device;
        type = "disk";
        content = {
          type = "table";
          format = "msdos";
          partitions = [
            {
              name = "root";
              part-type = "primary";
              start = "1M";
              end = "100%";
              bootable = true;
              content = {
                type = "filesystem";
                format = "ext4";
                mountpoint = "/";
              };
            }
          ];
        };
      };
    };
  };

  boot = {
    kernelPackages = pkgs.linuxPackages_latest;
    loader.grub = {
      enable = true;
      devices = [ device ];
    };
  };
```

And it won't boot:

![image](https://github.com/numtide/nixos-anywhere/assets/10998835/d4808114-b019-4be3-951f-5b135c053dfc)

Here's a a `fdisk -l` from the rescue mode:

![image](https://github.com/numtide/nixos-anywhere/assets/10998835/9dcbc5c2-12f6-4b5f-ac6f-6a039457c5b1)

**Comment by doronbehar** [comment]
Here is the generated fstab:

![image](https://github.com/numtide/nixos-anywhere/assets/10998835/d72910d6-3ddf-4ec5-9850-fe84998fe809)

**Comment by phaer** [comment]
Don't see why your kernel wouldn't find this block device, what kind of machine is that?

Does it maybe use nvme? I believe you'd need kernel modules for that. You could e.g. try with   `boot.initrd.availableKernelModules = ["xhci_pci" "ahci" "sd_mod" "nvme" ];` 
Or some exotic raid hardware or anything?

**Comment by doronbehar** [comment]
> Don't see why your kernel wouldn't find this block device, what kind of machine is that?

The provider's website says: "SSD SATA". Is there a way to check that via a grub / outer rescue mode?

> Does it maybe use nvme?

It doesn't, according to the provider's ([ovh](https://www.ovhcloud.com/)) website.

> I believe you'd need kernel modules for that. You could e.g. try with `boot.initrd.availableKernelModules = ["xhci_pci" "ahci" "sd_mod" "nvme" ];`

That didn't help - still won't boot.

> I wasn't talking about the name along, but also different feature sets. More examples is only good if the are maintained and tested over time - something we seem to even struggle with in the README.

Perhaps if the examples weren't embedded in the README, but linked by it (or embedded via [asciidoc's `include` feature](https://docs.asciidoctor.org/asciidoc/latest/verbatim/source-blocks/#using-include-directives-in-source-blocks)).

**Comment by doronbehar** [comment]
Thanks for your dedicated help in the meantime, I appreciate it!

**Comment by doronbehar** [comment]
BTW this is the grub command that was generated (maybe it'll help):

![image](https://github.com/numtide/nixos-anywhere/assets/10998835/e1792f07-60b6-4404-8785-2b8d60504eef)

**Comment by Lassulus** [comment]
can you add `boot.shell_on_fail` to your cmdline and check what's in /dev in the rescue shell? or maybe the output of `fdisk -l`. It seems to find the kernel but can't find the root partitions for some reason, so the problem seems to be in the initrd

**Comment by doronbehar** [comment]
Thanks for joining the discussion @Lassulus . Unfortunately, I had to make that server functional fast, and I eventually used [nixos-infect](https://github.com/elitak/nixos-infect). Hence I don't have access to that broken vps anymore.

I will note though, that even with `nixos-infect`, I experienced a disk recognition boot failure only when I chose the Latest Ubuntu 23.04 distro by ovh.. See https://github.com/elitak/nixos-infect/pull/155 . Hence this could be a different "hardware" that ovh sets up when it reinstalls your distro.

I won't have time in the near future to contribute to the docs as discussed earlier, but I'll certainly put it on my TODO!

**Comment by phaer** [comment]
> can you add boot.shell_on_fail to your cmdline and check what's in /dev in the rescue shell? or maybe the output of fdisk -l. It seems to find the kernel but can't find the root partitions for some reason, so the problem seems to be in the initrd

Yeah, could still be just a missing kernel module, hard to tell. Would be interesting to compare the list of loaded drivers with the shipped ubuntu 23.04 :) 

But given that @doronbehar doesn't have time for that, I'll close for now. Feel free to re-open if there's something new here :)
